README
NAME
    Net::Disqus - Disqus.com API access

VERSION
    version 1.18

SYNOPSIS
        use Net::Disqus;
        my $disqus = Net::Disqus->new(
            api_secret  => 'your_api_secret',
            %options,
            );

        my $reactions = $disqus->reactions->list(...);

OBJECT METHODS
  new(%options)
    Creates a new Net::Disqus object. Arguments that can be passed to the
    constructor:

        api_secret      (REQUIRED)  Your Disqus API secret
        secure          (optional)  When set, will use HTTPS instead of HTTP
        ua_args         (optional)  Arguments to pass to the user agent (See L<Net::Disqus::UserAgent>)
        pass_api_errors (optional)  If set, API errors are passed back as result so you can inspect it,
                                    otherwise an exception is thrown if an API error occurs.

  rate_limit
    Returns the rate limit you have on the Disqus API.

  rate_limit_remaining
    Returns the number of requests you have remaining out of your rate
    limit.

  rate_limit_reset
    Returns an epoch time when your rate limit will be reset.

  rate_limit_resets_in
    Returns the number of seconds until your rate limit will be reset.

  rate_limit_wait
    Returns the number of seconds you have to wait after the last performed
    request to exactly use up your rate limit before it's reset. Will return
    undef if the rate limit values aren't set, or for some reason you happen
    to hit it right when your limit resets. WARNING: Do not use
    sleep($disqus->rate_limit_wait), the net effect of sleep(undef) is that
    your application will sleep forever.

        An example:

            while(1) {
                my $reactions = $disqus->reactions->list(forum => 'mysite');
                my $wait = $diqus->rate_limit_wait;
                ...
                sleep($wait || 60); 
            }

  fetch(url, %args)
    Returns the result from an API call. Pass the following arguments:

        url     (REQUIRED)  The url to call on the Disqus API
        %args   (optional)  Arguments to pass to the API

    This method will use the interfaces.json file to check whether the
    endpoint you requested is valid and whether you have passed all required
    arguments. An exception is thrown for invalid endpoints or missing
    required arguments.

CALLING API METHODS
    You can call API methods either by their full URL:

        $disqus->fetch('/reactions/list', forum => 'foo')

    Or you can get the same result like this:

        $disqus->reactions->list(forum => 'foo');

    Use whatever you're more comfortable with.

    For a list of API methods and their arguments, please see
    <http://disqus.com/api/docs/>.

    All API calls will return a hash reference containing at the very least
    a 'code' and a 'response' key. Use the API documentation or API console
    to find out what exactly is returned.

EXCEPTION HANDLING
    When errors occur, and the 'pass_api_errors' option is not set,
    Net::Disqus will die and throw a Net::Disqus::Exception object. The
    object contains two properties, 'code' and 'text. Use the below table to
    find out what the problem was:

        Code    Text
        -------------------------------------------------------------------------------
        0       Success
        1       Endpoint not valid
        2       Missing or invalid argument
        3       Endpoint resource not valid
        4       You must provide an authenticated user for this method
        5       Invalid API key
        6       Invalid API version
        7       You cannot access this resource using %(request_method)s
        8       A requested object was not found
        9       You cannot access this resource using your %(access_type)s API key
        10      This operation is not supported
        11      Your API key is not valid on this domain
        12      This application does not have enough privileges to access this resource
        13      You have exceeded the rate limit for this resource
        14      You have exceeded the rate limit for your account
        15      There was internal server error while processing your request
        16      Your request timed out
        17      The authenticated user does not have access to this feature
        500*    No such API endpoint
        500*    <variable>

    The above list was taken from <http://disqus.com/api/docs/errors/>. The
    HTTP status codes are not returned in the exception object. Exception
    code '500' is Net::Disqus' own, and will either contain the 'No such API
    endpoint' message if you are trying to access an endpoint not defined,
    or whatever error LWP::UserAgent encountered.

    If you have set the 'pass_api_errors' option, the JSON error object will
    be returned to your application, so you will have to check the code
    yourself according to the above table.

EXCEPTION HANDLING EXAMPLES
    An example using Try::Tiny:

        my $nd = Net::Disqus->new(api_secret => 'reallyinvalid');
        my $res;

        try {
            $res = $nd->reactions->list(forum => 'foo');
        } catch {
            # $_ contains the Net::Disqus::Exception object
            #
            # $_->code should contain '5' and $_->text should contain 'Invalid API Key'
        };

    Another example:

        my $nd = Net::Disqus->new(api_secret => 'myrealapikeygoeshere');
        my $res;

        try {
            $res = $nd->fetch('/this/isnt/good');
        } catch {
            # $_ contains the Net::Disqus::Exception object
            #
            # $_->code should contain '500' and $_->text should contain 'No such API endpoint'
        };

    And an example where we want to do this ourselves:

        my $nd = Net::Disqus->new(api_secret => 'reallyinvalid', pass_api_errors => 1);
        my $res;

        try {
            $res = $nd->reactions->list(forum => 'foo');
        } catch {
            # this will not result in an error unless there was no connection possible
        };
        # and now
        # $res->{code} == 5;
        # $res->{response} eq 'Invalid API Key';

AUTHOR
    Ben van Staveren, "<madcat at cpan.org>"

BUGS AND/OR CONTRIBUTING
    Please report any bugs or feature requests through the web interface at
    <https://github.com/benvanstaveren/net-disqus/issues>. If you want to
    contribute code or patches, feel free to fork the Git repository located
    at <https://github.com/benvanstaveren/net-disqus> and make pull requests
    for any patches you have.

SUPPORT
    You can find documentation for this module with the perldoc command.

        perldoc Net::Disqus

    You can also look for information at:

    *   AnnoCPAN: Annotated CPAN documentation

        <http://annocpan.org/dist/Net-Disqus>

    *   CPAN Ratings

        <http://cpanratings.perl.org/d/Net-Disqus>

    *   Search CPAN

        <http://search.cpan.org/dist/Net-Disqus/>

ACKNOWLEDGEMENTS
    Parts of the code based on Disqus' official Python lib.

LICENSE AND COPYRIGHT
    Copyright 2011 Ben van Staveren.

    This program is free software; you can redistribute it and/or modify it
    under the terms of either: the GNU General Public License as published
    by the Free Software Foundation; or the Artistic License.

    See http://dev.perl.org/licenses/ for more information.



Changes
Revision history for Net-Disqus

1.18      2011-04-28 11:48:07 Asia/Jakarta
    Fixed missing prerequisites 

1.17      2011-04-25 05:39:55 Asia/Jakarta
    Fix for the Interfaces.pm module, all is in order now.
    

1.16      2011-04-25 04:39:25 Asia/Jakarta
    The interface.json file got removed in favor as including it as a module


Build.PL

use strict;
use warnings;

use Module::Build 0.38;


my %module_build_args = (
  'build_requires' => {
    'Module::Build' => '0.38'
  },
  'configure_requires' => {
    'ExtUtils::MakeMaker' => '6.31',
    'Module::Build' => '0.38'
  },
  'dist_abstract' => 'Disqus.com API access',
  'dist_author' => [
    'Ben van Staveren <madcat@cpan.org>'
  ],
  'dist_name' => 'Net-Disqus',
  'dist_version' => '1.18',
  'license' => 'perl',
  'module_name' => 'Net::Disqus',
  'recommends' => {},
  'recursive_test_files' => 1,
  'requires' => {
    'Class::Accessor' => '0',
    'JSON::PP' => '0',
    'LWP::UserAgent' => '0',
    'Try::Tiny' => '0',
    'URI' => '0',
    'URI::Escape' => '0',
    'overload' => '0'
  },
  'script_files' => []
);


my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


Makefile.PL

use strict;
use warnings;



use ExtUtils::MakeMaker 6.31;



my %WriteMakefileArgs = (
  'ABSTRACT' => 'Disqus.com API access',
  'AUTHOR' => 'Ben van Staveren <madcat@cpan.org>',
  'BUILD_REQUIRES' => {
    'Module::Build' => '0.38'
  },
  'CONFIGURE_REQUIRES' => {
    'ExtUtils::MakeMaker' => '6.31',
    'Module::Build' => '0.38'
  },
  'DISTNAME' => 'Net-Disqus',
  'EXE_FILES' => [],
  'LICENSE' => 'perl',
  'NAME' => 'Net::Disqus',
  'PREREQ_PM' => {
    'Class::Accessor' => '0',
    'JSON::PP' => '0',
    'LWP::UserAgent' => '0',
    'Try::Tiny' => '0',
    'URI' => '0',
    'URI::Escape' => '0',
    'overload' => '0'
  },
  'VERSION' => '1.18',
  'test' => {
    'TESTS' => 't/*.t'
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/Net/Disqus/UserAgent.pm
use strict;
use warnings;
package Net::Disqus::UserAgent;
BEGIN {
  $Net::Disqus::UserAgent::VERSION = '1.18';
}
use Net::Disqus::Exception;
use Try::Tiny;

sub new {
    my $pkg = shift;
    my %args = (
        pass_content_as_is  => 0,
        forcelwp            => 0,
        @_,
        ua_class            => 'LWP::UserAgent',
        ua_key              => 'lwp',
    );

    $args{agent} ||= "Net::Disqus/$Net::Disqus::VERSION";
    if(!$args{'forcelwp'}) {
        eval 'use Mojo::UserAgent; use Mojo::JSON; use Mojo::URL';
        unless($@) {
            $args{'ua_class'} = 'Mojo::UserAgent';
            $args{'name'} = delete($args{'agent'});
            $args{'ua_key'} = 'mojo';
        }  else {
            eval 'use LWP::UserAgent; use JSON::PP; use URI; use URI::Escape;';
            die Net::Disqus::Exception->new({code => 500, text => 'Something really funny is going on, cannot find one of LWP::UserAgent, JSON::PP, URI, URI::Escape'}) if($@);
        }
    } else {
        eval 'use LWP::UserAgent; use JSON::PP; use URI; use URI::Escape;';
        die Net::Disqus::Exception->new({code => 500, text => 'Something really funny is going on, cannot find one of LWP::UserAgent, JSON::PP, URI, URI::Escape'}) if($@);
    }
    my $self = bless({%args}, $pkg);
    delete($args{$_}) for(qw(pass_content_as_is forcelwp ua_class ua_key)); # and this is for LWP who doesn't like being passed unknown options
    $self->{'ua'} = $self->{'ua_class'}->new(%args);
    return $self;
}

sub ua { return shift->{ua} }
sub ua_key { return shift->{ua_key} }
sub ua_class { return shift->{ua_class} }
sub pass_content_as_is { return shift->{pass_content_as_is} }

sub request { 
    my $self = shift; 
    my $method = shift; 
    my $f = "tx_" . $self->ua_key; 
    return $self->$f($method, @_);
};

sub tx_mojo {
    my $self = shift;
    my $method = shift;
    my $url = shift;
    my %args = (@_);
    my $rate = {};

    my $uri = Mojo::URL->new(
        ($method eq 'get') 
            ?  sprintf('%s?%s', $url, join('&', map { sprintf('%s=%s', $_, $args{$_}) } (keys(%args))))
            : $url
        );
    my $f = ($method eq 'get') 
        ? 'get' 
        : 'post_form';
    my @fa = ($uri);
    push(@fa, { %args }) if($method eq 'post');

    my $res = $self->ua->$f(@fa)->res;
    die Net::Disqus::Exception->new({ code => 500, text => 'Did not receive a JSON response'}) if( 
        ($res->headers->content_type && $res->headers->content_type ne 'application/json') && 
        !$self->pass_content_as_is
        );

    $rate->{$_} = $res->headers->to_hash->{$_} || 0 for(qw(X-Ratelimit-Remaining X-Ratelimit-Limit X-Ratelimit-Reset));
    my @ret = (
        ($self->pass_content_as_is) ? $res->body : $res->json, 
        $rate
    );
    return @ret;
}

sub json_decode {
    my $self = shift;
    my $str  = shift;

    return ($self->ua_key eq 'mojo') 
        ? Mojo::JSON->decode($str)
        : JSON::PP::decode_json($str);
}

sub tx_lwp {
    my $self = shift;
    my $method = shift;
    my $url  = shift;
    my %args = (@_);
    my $rate = {};

    my $uri = URI->new($url);
    my $query_args = join('&', map { sprintf('%s=%s', $_, uri_escape($args{$_})) } (keys(%args)));
    $uri->query($query_args) if($method eq 'get');

    my $request = HTTP::Request->new(uc($method), $uri);
    $request->content($query_args) if($method eq 'post');
    my $res = $self->ua->request($request);
    die Net::Disqus::Exception->new({ code => 500, text => 'Did not receive a JSON response'}) if($res->header('Content-Type') ne 'application/json' && !$self->pass_content_as_is);
    my $json;
    if($self->pass_content_as_is) {
        $json = $res->content;
    } else {
        try {
            $json = JSON::PP::decode_json($res->content);
        } catch {
            die Net::Disqus::Exception->new({ code => 500, text => "Failed JSON decoding: $_"});
        };
    }
    $rate->{$_} = $res->header($_) || 0 for(qw(X-Ratelimit-Remaining X-Ratelimit-Limit X-Ratelimit-Reset));
    return ($json, $rate);
}

1;

__END__
=head1 NAME

Net::Disqus::UserAgent - Wrapper around LWP::UserAgent or Mojo::UserAgent

=head1 VERSION

version 1.18

=head1 SYNOPSIS
    
    # Do not use this module directly, it's full of little internal tidbits for
    # Net::Disqus, this is just here as an example

    use Net::Disqus::UserAgent
    my $ua = Net::Disqus::UserAgent->new(%options);

=head1 OBJECT METHODS

=head2 new(%options)
    
Creates a new Net::Disqus::UserAgent object. This is usually done by L<Net::Disqus>, but the options below are valid to pass to the 'ua_args' option in the constructor for L<Net::Disqus>.

    forcelwp            (optional)  When set to a true value, will always use LWP::UserAgent even if Mojo::UserAgent is available
    pass_content_as_is  (optional)  When set, will not check to see whether a JSON response was returned, and will not attempt any decoding, but will return content as-is.

=head1 USER AGENT AUTO DETECTION

If you don't force LWP, Net::Disqus::UserAgent will try in order:

    Mojo::UserAgent
    LWP::UserAgent

LWP::UserAgent is a requirement for Net::Disqus to be installed, so at the very least you will always have that. This behaviour was introduced for the L<Mojolicious::Plugin::Disqus> plugin, to make sure that we always use the best user agent for a given job.

=head1 AUTHOR

Ben van Staveren, C<< <madcat at cpan.org> >>

=head1 SEE ALSO

L<Net::Disqus>

=head1 LICENSE AND COPYRIGHT

Copyright 2011 Ben van Staveren.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=cut

lib/Net/Disqus/Interfaces.pm
use strict;
use warnings;
package Net::Disqus::Interfaces;
BEGIN {
  $Net::Disqus::Interfaces::VERSION = '1.18';
}

# this got turned into a module since bundling a json file is often
# a pain in the you-know-whatsit - plus I forgot to add it to the Dist::Zilla
# distribution last time, so as a module at least it get picked up all the time 
# regardless. 
sub INTERFACES { return {'imports' => {'details' => {'formats' => ['json','jsonp'],'required' => ['group'],'method' => 'GET'},'list' => {'formats' => ['json','jsonp'],'required' => ['forum'],'method' => 'GET'}},'applications' => {'listUsage' => {'formats' => ['json','jsonp'],'required' => [],'method' => 'GET'}},'categories' => {'listPosts' => {'formats' => ['json','jsonp','rss'],'required' => ['category'],'method' => 'GET'},'details' => {'formats' => ['json','jsonp'],'required' => ['category'],'method' => 'GET'},'list' => {'formats' => ['json','jsonp'],'required' => [],'method' => 'GET'},'listThreads' => {'formats' => ['json','jsonp','rss'],'required' => ['category'],'method' => 'GET'}},'users' => {'follow' => {'formats' => ['json','jsonp'],'required' => [],'method' => 'POST'},'listFollowers' => {'formats' => ['json','jsonp'],'required' => [],'method' => 'GET'},'listMostActiveForums' => {'formats' => ['json','jsonp'],'required' => [],'method' => 'GET'},'listForums' => {'formats' => ['json','jsonp'],'required' => [],'method' => 'GET'},'listActivity' => {'formats' => ['json','jsonp'],'required' => [],'method' => 'GET'},'details' => {'formats' => ['json','jsonp'],'required' => [],'method' => 'GET'},'unfollow' => {'formats' => ['json','jsonp'],'required' => [],'method' => 'POST'},'listFollowing' => {'formats' => ['json','jsonp'],'required' => [],'method' => 'GET'},'listActiveForums' => {'formats' => ['json','jsonp'],'required' => [],'method' => 'GET'}},'threads' => {'create' => {'formats' => ['json','jsonp'],'required' => ['forum','title'],'method' => 'POST'},'remove' => {'formats' => ['json','jsonp'],'required' => ['thread'],'method' => 'POST'},'listSimilar' => {'formats' => ['json','jsonp','rss'],'required' => ['thread'],'method' => 'GET'},'listByDate' => {'formats' => ['json','jsonp','rss'],'required' => [],'method' => 'GET'},'vote' => {'formats' => ['json','jsonp'],'required' => ['thread','vote'],'method' => 'POST'},'restore' => {'formats' => ['json','jsonp'],'required' => ['thread'],'method' => 'POST'},'details' => {'formats' => ['json','jsonp'],'required' => ['thread'],'method' => 'GET'},'listHot' => {'formats' => ['json','jsonp','rss'],'required' => [],'method' => 'GET'},'open' => {'formats' => ['json','jsonp'],'required' => ['thread'],'method' => 'POST'},'listMostLiked' => {'formats' => ['json','jsonp','rss'],'required' => [],'method' => 'GET'},'close' => {'formats' => ['json','jsonp'],'required' => ['thread'],'method' => 'POST'},'listPosts' => {'formats' => ['json','jsonp','rss'],'required' => ['thread'],'method' => 'GET'},'update' => {'formats' => ['json','jsonp'],'required' => ['thread'],'method' => 'POST'},'list' => {'formats' => ['json','jsonp','rss'],'required' => [],'method' => 'GET'}},'whitelists' => {'remove' => {'formats' => ['json','jsonp'],'required' => ['forum'],'method' => 'POST'},'add' => {'formats' => ['json','jsonp'],'required' => ['forum'],'method' => 'POST'},'list' => {'formats' => ['json','jsonp'],'required' => ['forum'],'method' => 'GET'}},'exports' => {'exportForum' => {'formats' => ['json','jsonp'],'required' => ['forum'],'method' => 'POST'}},'reactions' => {'details' => {'formats' => ['json','jsonp'],'required' => ['reaction'],'method' => 'GET'},'list' => {'formats' => ['json','jsonp'],'required' => ['forum'],'method' => 'POST'}},'blacklists' => {'remove' => {'formats' => ['json','jsonp'],'required' => ['forum'],'method' => 'POST'},'add' => {'formats' => ['json','jsonp'],'required' => ['forum'],'method' => 'POST'},'list' => {'formats' => ['json','jsonp'],'required' => ['forum'],'method' => 'GET'}},'trends' => {'listThreads' => {'formats' => ['json','jsonp','rss'],'required' => [],'method' => 'GET'}},'forums' => {'listUsers' => {'formats' => ['json','jsonp'],'required' => ['forum'],'method' => 'GET'},'listPosts' => {'formats' => ['json','jsonp','rss'],'required' => ['forum'],'method' => 'GET'},'listMostLikedUsers' => {'formats' => ['json','jsonp'],'required' => ['forum'],'method' => 'GET'},'details' => {'formats' => ['json','jsonp'],'required' => ['forum'],'method' => 'GET'},'listCategories' => {'formats' => ['json','jsonp'],'required' => ['forum'],'method' => 'GET'},'listThreads' => {'formats' => ['json','jsonp','rss'],'required' => ['forum'],'method' => 'GET'}},'reports' => {'domains' => {'formats' => ['json','jsonp'],'required' => [],'method' => 'GET'},'ips' => {'formats' => ['json','jsonp'],'required' => [],'method' => 'GET'},'users' => {'formats' => ['json','jsonp'],'required' => [],'method' => 'GET'},'threads' => {'formats' => ['json','jsonp'],'required' => [],'method' => 'GET'}},'posts' => {'create' => {'formats' => ['json','jsonp'],'required' => ['message'],'method' => 'POST'},'remove' => {'formats' => ['json','jsonp'],'required' => ['post'],'method' => 'POST'},'search' => {'formats' => ['json','jsonp','rss'],'required' => ['query'],'method' => 'GET'},'vote' => {'formats' => ['json','jsonp'],'required' => ['vote','post'],'method' => 'POST'},'restore' => {'formats' => ['json','jsonp'],'required' => ['post'],'method' => 'POST'},'details' => {'formats' => ['json','jsonp'],'required' => ['post'],'method' => 'GET'},'approve' => {'formats' => ['json','jsonp'],'required' => ['post'],'method' => 'POST'},'spam' => {'formats' => ['json','jsonp'],'required' => ['post'],'method' => 'POST'},'report' => {'formats' => ['json','jsonp'],'required' => ['post'],'method' => 'POST'},'listPopular' => {'formats' => ['json','jsonp','rss'],'required' => [],'method' => 'GET'},'list' => {'formats' => ['json','jsonp','rss'],'required' => [],'method' => 'GET'}}}; }

1;


lib/Net/Disqus.pm
use warnings;
use strict;
package Net::Disqus;
BEGIN {
  $Net::Disqus::VERSION = '1.18';
}
use Try::Tiny;
use Net::Disqus::UserAgent;
use Net::Disqus::Interfaces;
use Net::Disqus::Exception;
use base 'Class::Accessor';

__PACKAGE__->mk_ro_accessors(qw(api_key api_secret api_url ua pass_api_errors));
__PACKAGE__->mk_accessors(qw(interfaces rate_limit rate_limit_remaining rate_limit_reset fragment path));

our $AUTOLOAD;

sub new {
    my $class = shift;

    die Net::Disqus::Exception->new({ code => 500, text => '"new" is not an instance method'}) if(ref($class));

    my %args = (
        secure => 0,
        pass_api_errors => 0,
        ua_args => {},
        (@_ == 1 && ref $_[0] eq 'HASH') ? %{$_[0]} : @_,
        interfaces => {},
        api_url => 'http://disqus.com/api/3.0',
        );

    die Net::Disqus::Exception->new({ code => 500, text => "missing required argument 'api_secret'"}) unless $args{'api_secret'};

    $args{'ua'} = Net::Disqus::UserAgent->new(%{$args{'ua_args'}});
    $args{'api_url'} = 'https://secure.disqus.com/api/3.0' if($args{'secure'});
    my $self = $class->SUPER::new({%args});

    $self->interfaces(Net::Disqus::Interfaces->INTERFACES());
    return $self;
}

sub fetch {
    my $self = shift;
    my $url  = shift;
    my $t = $self;

    $url =~ s/^\///;
    my @url = split(/\//, $url);
    my $last = pop(@url);

    $t = $t->$_() for(@url);
    return $t->$last(@_);
}

sub rate_limit_resets_in {
    my $self = shift;
    my $now = time();

    return ($now > $self->rate_limit_reset) 
        ? undef 
        : $self->rate_limit_reset - $now;
}

sub rate_limit_wait {   
    my $self = shift;
    my $now = time();
    my $reset = $self->rate_limit_reset || 0;

    return undef unless($reset > 0); 
    return undef if($now > $reset);

    my $diff = $reset - $now;
    my $remaining = $self->rate_limit_remaining;

    return undef if($diff == 0 || $remaining == 0);
    
    # we can do X requests every Y seconds to fill it up right
    # to the reset time
    my $wait = int($diff/$remaining);
    $wait-- if($wait * $remaining > $diff); 
    return $wait;
}

sub _mk_request {
    my $self = shift;
    my $fragment = $self->fragment;
    my %args = (@_);

    $self->fragment(undef);

    my $url = sprintf('%s%s.json', $self->api_url, $self->path);
    my $method = lc($fragment->{method});
    my $required = $fragment->{required} || [];

    for(@$required) {
        die Net::Disqus::Exception->new({ code => 500, text => "missing required argument '$_'"}) unless($args{$_});
    }
    $args{'api_secret'} = $self->api_secret;

    # and there we are. 
    my ($json, $rate) = $self->ua->request($method, $url, %args);
    die Net::Disqus::Exception->new({ code => $json->{code}, text => $json->{response}}) if(!$self->pass_api_errors && $json->{code} != 0);

    $self->rate_limit($rate->{'X-Ratelimit-Limit'});
    $self->rate_limit_remaining($rate->{'X-Ratelimit-Remaining'});
    $self->rate_limit_reset($rate->{'X-Ratelimit-Reset'});
    return $json;
}

sub AUTOLOAD {
    my $self = shift;
    my $fragment = ((($_ = $AUTOLOAD) =~ s/.*://) ? $_ : "");
    return if($fragment eq uc($fragment));

    unless($self->fragment) {
        $self->fragment($self->interfaces);
        $self->path('');
    }
    $self->path($self->path . '/' . $fragment);
    if($self->fragment->{$fragment}) {
        $self->fragment($self->fragment->{$fragment});
        return ($self->fragment->{method}) 
            ? $self->_mk_request(@_)
            : $self;
    } else {
        $self->fragment(undef);
        my $path = $self->path and $self->path(undef);
        die Net::Disqus::Exception->new({ code => 500, text => "No such API endpoint"});
    }
}
        
1;
__END__
=head1 NAME

Net::Disqus - Disqus.com API access

=head1 VERSION

version 1.18

=head1 SYNOPSIS

    use Net::Disqus;
    my $disqus = Net::Disqus->new(
        api_secret  => 'your_api_secret',
        %options,
        );

    my $reactions = $disqus->reactions->list(...);

=head1 OBJECT METHODS

=head2 new(%options)
    
Creates a new Net::Disqus object. Arguments that can be passed to the constructor:

    api_secret      (REQUIRED)  Your Disqus API secret
    secure          (optional)  When set, will use HTTPS instead of HTTP
    ua_args         (optional)  Arguments to pass to the user agent (See L<Net::Disqus::UserAgent>)
    pass_api_errors (optional)  If set, API errors are passed back as result so you can inspect it,
                                otherwise an exception is thrown if an API error occurs.

=head2 rate_limit

Returns the rate limit you have on the Disqus API.

=head2 rate_limit_remaining

Returns the number of requests you have remaining out of your rate limit.

=head2 rate_limit_reset

Returns an epoch time when your rate limit will be reset.

=head2 rate_limit_resets_in

Returns the number of seconds until your rate limit will be reset.

=head2 rate_limit_wait

Returns the number of seconds you have to wait after the last performed request to exactly use up your rate limit before it's reset. Will return undef if the rate limit values aren't set,
or for some reason you happen to hit it right when your limit resets. WARNING: Do not use sleep($disqus->rate_limit_wait), the net effect of sleep(undef) is that your application will
sleep forever.

    An example:

        while(1) {
            my $reactions = $disqus->reactions->list(forum => 'mysite');
            my $wait = $diqus->rate_limit_wait;
            ...
            sleep($wait || 60); 
        }

=head2 fetch(url, %args)
   
Returns the result from an API call. Pass the following arguments:

    url     (REQUIRED)  The url to call on the Disqus API
    %args   (optional)  Arguments to pass to the API

This method will use the interfaces.json file to check whether the endpoint you requested is valid and whether you have passed all required arguments. An exception is thrown for invalid endpoints or missing required arguments.


=head1 CALLING API METHODS

You can call API methods either by their full URL:

    $disqus->fetch('/reactions/list', forum => 'foo')
    
Or you can get the same result like this:

    $disqus->reactions->list(forum => 'foo');

Use whatever you're more comfortable with.

For a list of API methods and their arguments, please see L<http://disqus.com/api/docs/>. 

All API calls will return a hash reference containing at the very least a 'code' and a 'response' key. Use the API documentation or API console to find out what exactly is returned.

=head1 EXCEPTION HANDLING

When errors occur, and the 'pass_api_errors' option is not set, Net::Disqus will die and throw a Net::Disqus::Exception object. The object contains two properties, 'code' and 'text. Use the below table to find out what the problem was:

    Code    Text
    -------------------------------------------------------------------------------
    0       Success
    1       Endpoint not valid
    2       Missing or invalid argument
    3       Endpoint resource not valid
    4       You must provide an authenticated user for this method
    5       Invalid API key
    6       Invalid API version
    7       You cannot access this resource using %(request_method)s
    8       A requested object was not found
    9       You cannot access this resource using your %(access_type)s API key
    10      This operation is not supported
    11      Your API key is not valid on this domain
    12      This application does not have enough privileges to access this resource
    13      You have exceeded the rate limit for this resource
    14      You have exceeded the rate limit for your account
    15      There was internal server error while processing your request
    16      Your request timed out
    17      The authenticated user does not have access to this feature
    500*    No such API endpoint
    500*    <variable>

The above list was taken from L<http://disqus.com/api/docs/errors/>. The HTTP status codes are not returned in the exception object. Exception code '500' is Net::Disqus' own, and will
either contain the 'No such API endpoint' message if you are trying to access an endpoint not defined, or whatever error LWP::UserAgent encountered.

If you have set the 'pass_api_errors' option, the JSON error object will be returned to your application, so you will have to check the code yourself according to the above table.

=head1 EXCEPTION HANDLING EXAMPLES

An example using Try::Tiny:

    my $nd = Net::Disqus->new(api_secret => 'reallyinvalid');
    my $res;

    try {
        $res = $nd->reactions->list(forum => 'foo');
    } catch {
        # $_ contains the Net::Disqus::Exception object
        #
        # $_->code should contain '5' and $_->text should contain 'Invalid API Key'
    };

Another example:

    my $nd = Net::Disqus->new(api_secret => 'myrealapikeygoeshere');
    my $res;

    try {
        $res = $nd->fetch('/this/isnt/good');
    } catch {
        # $_ contains the Net::Disqus::Exception object
        #
        # $_->code should contain '500' and $_->text should contain 'No such API endpoint'
    };

And an example where we want to do this ourselves:

    my $nd = Net::Disqus->new(api_secret => 'reallyinvalid', pass_api_errors => 1);
    my $res;

    try {
        $res = $nd->reactions->list(forum => 'foo');
    } catch {
        # this will not result in an error unless there was no connection possible
    };
    # and now
    # $res->{code} == 5;
    # $res->{response} eq 'Invalid API Key';

=head1 AUTHOR

Ben van Staveren, C<< <madcat at cpan.org> >>

=head1 BUGS AND/OR CONTRIBUTING

Please report any bugs or feature requests through the web interface at L<https://github.com/benvanstaveren/net-disqus/issues>. If you want to contribute code or patches, feel free to fork the Git repository located at L<https://github.com/benvanstaveren/net-disqus> and make pull requests for any patches you have.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Net::Disqus


You can also look for information at:

=over 4

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Net-Disqus>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Net-Disqus>

=item * Search CPAN

L<http://search.cpan.org/dist/Net-Disqus/>

=back

=head1 ACKNOWLEDGEMENTS

Parts of the code based on Disqus' official Python lib.

=head1 LICENSE AND COPYRIGHT

Copyright 2011 Ben van Staveren.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=cut